### ISSUE TITLE
Ports

### ISSUE THREAD

**Issue by shakna-israel** [shakna-israel/f.lua#39]
Ports instead of files.

Basic string-like objects that can be written to, which can then write to either files or actual ports. Can also be read from.

**Comment by shakna-israel** [comment]
Reading:

```port(1, 'r', "*all")```

```port(input, 'r', '*all')```

```port(output, 'w', "*all")```

Third var refers to an in-memory piece of data, which gets cleared after it is read in it's entirety. (e.g. EOF is found by any port() call.)

**Comment by shakna-israel** [comment]
Any port read from, changes to that particular port's data.

e.g.

```
port(1, 'w', "Hello, World!")
port(1, 'r', "*all")
> "Hello, World!"
```

**Comment by shakna-israel** [comment]
What about something like:

```
port(number, mode, string-data)
port(filename, mode, string-data)
port(number, mode, port-number)
```

So writing to a file called tmp, from stdin could look like:

```
port("tmp", "w", 0)
```

**Comment by shakna-israel** [comment]
A simple echo would look like:

```
port(1, "w", 0)
```

An expanded form might be:

```
port(1, "w", port(0, "r", 0))
```

So a little bit of magic, but not much, and in a way that should be somewhat familiar to Lispers.

**Comment by shakna-israel** [comment]
I'm working on something based more closely around [Scheme's port library](https://wiki.call-cc.org/man/4/Unit%20ports).

It should add up to seven functions, that either take or return a port.

Thinking:

```lua
-- rebind stdout to "port"
with-output-to-port(port, functor)

-- rebind stderr to "port"
with-error-output-to-port(port, functor)

-- rebind stdin to "port"
with-input-from-port(port, functor)

make-input-port(read-func, close-func)

make-output-port(write-func, close-func)

-- Constructs a port containing a string, and binds that to the port given to functor.
with-input-from-string(string, functor)

port-for (Iterator)
```

**Comment by shakna-israel** [comment]
Demo implementation seems to be coming along nicely:

```lua
local ports = require "ports"

local p = ports.make_output_port(function(self, data)
  self.data = data
end,
function(self)
  return self.data
end,
function(self)
  return nil
end)

ports.with_output_to_port(p, function()
  print("Hello, World!")
  io.write("Hi")
  io.write("Hi")
end)

print(p:read())
```

```
Hello, World!
HiHi
```

**Comment by shakna-israel** [comment]
Implemented in 95611d0d4f2d128be7ce13b568b1fbed230257e2.
